server/controller/src/main/java/ai/starwhale/mlops/domain/model/ModelService.java

```diff
@@ -622,9 +659,62 @@ public void uploadSrc(Long modelVersionId, MultipartFile multipartFile, ModelUpl
         modelVersionMapper.updateStatus(modelVersionId, ModelVersionEntity.STATUS_AVAILABLE);
     }
 
+    private boolean syncBuiltInRuntime(Long modelVersionId) {
+        // check whether it have a built-in runtime
+        ModelVersionEntity modelVersionEntity = modelDao.getModelVersion(modelVersionId.toString());
+        if (modelVersionEntity == null) {
+            log.warn("built-in runtime upload error, no model version found");
+            return false;
+        }
+        try {
+            var model = modelDao.getModel(modelVersionEntity.getModelId());
+            String manifest = getManifest(modelVersionEntity);
+            var metaInfo = Constants.yamlMapper.readValue(manifest, MetaInfo.class);
+            if (metaInfo != null && metaInfo.getPackagedRuntime() != null) {
+                var runtime = metaInfo.getPackagedRuntime();
+
+                var runtimeRequest = new ClientRuntimeRequest();
+                runtimeRequest.setRuntime(runtime.getName() + ":" + runtime.getManifest().getVersion());
+                runtimeRequest.setProject(model.getProjectId().toString());
+                runtimeRequest.setForce("0");
+                var resourceOptional = metaInfo.getResources().stream()
+                        .filter(r -> r.getPath().equalsIgnoreCase(runtime.getPath()))
+                        .findFirst();
+                if (resourceOptional.isPresent()) {
+                    var resource = resourceOptional.get();
+                    String filePath;
+                    switch (resource.getDesc()) {
+                        case SRC:
+                            filePath = String.format(
+                                    FORMATTER_STORAGE_PATH, modelVersionEntity.getStoragePath(), runtime.getPath());
+                            break;
+                        case MODEL:
+                            var project = projectService.findProject(model.getProjectId());
+                            filePath = storagePathCoordinator.allocateCommonModelPoolPath(
+                                    project.getId(), runtime.getHash());
+                            break;
+                        default:
+                            throw new SwValidationException(
+                                    ValidSubject.MODEL, "unsupported type " + resource.getDesc());
+                    }
+                    var fileName = String.format("%s.swrt", runtime.getManifest().getVersion());
+                    runtimeService.upload(
+                        new CustomMultipartFile(fileName, storageAccessService.get(filePath)), runtimeRequest);
+                    return true;
+                } else {
+                    log.warn("built-in runtime upload error, no path found");
+                }
+            }
+        } catch (Exception e) {
+            log.warn("sync built-in runtime error for model version:{},error:{}", modelVersionEntity.getId(), e);
+        }
+        return false;
```

Comment:
Would it be better for the runtime to let the client upload a copy and tell the relationship?

Reply:
done